cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/program/MessagingProgramStateWriter.java

```diff
@@ -53,16 +53,30 @@
  */
 public final class MessagingProgramStateWriter implements ProgramStateWriter {
   private static final Logger LOG = LoggerFactory.getLogger(MessagingProgramStateWriter.class);
-  private static final Gson GSON = ApplicationSpecificationAdapter.addTypeAdapters(new GsonBuilder()).create();
-  private final MessagingService messagingService;
-  private final TopicId topicId;
-  private final RetryStrategy retryStrategy;
+  private static final Gson GSON =
+    ApplicationSpecificationAdapter.addTypeAdapters(new GsonBuilder())
+      .registerTypeAdapter(Arguments.class, new ArgumentsCodec())
+      .registerTypeAdapter(ProgramOptions.class, new ProgramOptionsCodec()).create();
+
+  private final ProgramStatePublisher programStatePublisher;
+
+  private ScheduledExecutorService scheduledExecutorService;
```

Comment:
This scheduling is at the wrong level, it should not be in this class. There is nothing in the StateWriter contract that forces you to use a separate instance per program run, but this implementation assumes there is one writer per run. For example, the state writer is also used in the CDAP master, where this logic will be wrong.

The twill application should have a scheduled service that uses the state writer, not the other way around.